src/meta/meta_turtle.c

```diff
@@ -218,19 +326,94 @@ __wt_turtle_exists(WT_SESSION_IMPL *session, bool *existp)
+ */
+static int
+__metadata_add_backup_target_uri(WT_SESSION_IMPL *session, const char *name, size_t len)
+{
+    WT_BACKUP_TARGET *new_target_uri;
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+    uint64_t bucket, hash;
+
+    conn = S2C(session);
+
+    WT_ERR(__wt_calloc_one(session, &new_target_uri));
+    WT_ERR(__wt_strndup(session, name, len, &new_target_uri->name));
+
+    hash = __wt_hash_city64(name, len);
+    bucket = hash & (conn->hash_size - 1);
+    new_target_uri->name_hash = hash;
+    /* Insert target uri entry into hashtable. */
+    TAILQ_INSERT_HEAD(&(conn)->backuphash[bucket], new_target_uri, hashq);
+
+    return (0);
+err:
+    if (new_target_uri != NULL)
+        __wt_free(session, new_target_uri->name);
+    __wt_free(session, new_target_uri);
+
+    return (ret);
+}
+
+/*
+ * __metadata_load_target_uri_list --
+ *     Load the list of target uris and construct a hashtable from it.
+ */
+static int
+__metadata_load_target_uri_list(WT_SESSION_IMPL *session, bool exist_backup, const char *cfg[])
+{
+    WT_CONFIG backup_config;
+    WT_CONFIG_ITEM cval, k, v;
+    WT_DECL_RET;
+
+    WT_TRET(__wt_config_gets(session, cfg, "backup_restore_target", &cval));
+    if (cval.len != 0) {
+        if (!exist_backup)
+            WT_RET_MSG(session, EINVAL,
+              "restoring a partial backup requires the WiredTiger metadata backup file.");
+        F_SET(S2C(session), WT_CONN_BACKUP_PARTIAL_RESTORE);
+
+        /*
+         * Check that the configuration string only has table schema formats in the target list and
+         * construct the target hash table.
+         */
+        __wt_config_subinit(session, &backup_config, &cval);
+        while ((ret = __wt_config_next(&backup_config, &k, &v)) == 0) {
+            if (!WT_PREFIX_MATCH(k.str, "table:"))
+                WT_RET_MSG(session, EINVAL,
+                  "partial backup restore only supports objects of type \"table\" formats in the "
+                  "target uri list, found %.*s instead.",
+                  (int)k.len, k.str);
+            WT_RET(__metadata_add_backup_target_uri(session, (char *)k.str, k.len));
+        }
+        WT_RET_NOTFOUND_OK(ret);
+    }
+    return (0);
+}
 /*
  * __wt_turtle_init --
  *     Check the turtle file and create if necessary.
  */
 int
-__wt_turtle_init(WT_SESSION_IMPL *session, bool verify_meta)
+__wt_turtle_init(WT_SESSION_IMPL *session, bool verify_meta, const char *cfg[])
 {
+    WT_BACKUP_TARGET *target_uri;
+    WT_CONNECTION_IMPL *conn;
     WT_DECL_RET;
+    uint64_t i;
     char *metaconf, *unused_value;
     bool exist_backup, exist_incr, exist_isrc, exist_turtle;
     bool load, load_turtle, validate_turtle;
 
+    conn = S2C(session);
     load = load_turtle = validate_turtle = false;
+    /* Initialize target uri hashtable. */
+    WT_ERR(__wt_calloc_def(session, conn->hash_size, &conn->backuphash));
+    for (i = 0; i < conn->hash_size; ++i)
+        TAILQ_INIT(&conn->backuphash[i]);
```

Comment:
Now that it is in here, the hash bucket could be a local variable that gets passed into the other functions. It is good it is allocated and freed only in here. I understand the convenience of having the connection available everywhere. We are single threaded here guaranteed, and the session is also available.

Reply:
Yes - I'm with Sue, that I'd keep the bucket local and pass it as a new parameter where it's needed.

Reply:
Done!